src/bt-hh-io.c

```diff
@@ -0,0 +1,714 @@
+
+  status = bthh_interface->set_protocol(&bd_addr, protocolMode);
+  if (status != BT_STATUS_SUCCESS)
+    goto err_bthh_interface_set_protocol;
+
+  init_pdu(&wbuf->buf.pdu, cmd->service, cmd->opcode);
+  send_pdu(wbuf);
+
+  return BT_STATUS_SUCCESS;
+err_bthh_interface_set_protocol:
+  cleanup_pdu_wbuf(wbuf);
+  return status;
+}
+
+static bt_status_t
+opcode_get_report(const struct pdu* cmd)
+{
+  long off;
+  bt_bdaddr_t bd_addr;
+  bthh_report_type_t reportType;
+  uint8_t reportId;
+  uint16_t bufferSize;
+  struct pdu_wbuf* wbuf;
+  bt_status_t status;
+
+  assert(bthh_interface);
+  assert(bthh_interface->get_report);
+
+  off = read_bt_bdaddr_t(cmd, 0, &bd_addr);
+  if (off < 0)
+    return BT_STATUS_PARM_INVALID;
+
+  off = read_pdu_at(cmd, off, "CCS", &reportType, &reportId, &bufferSize);
+  if (off < 0)
+    return BT_STATUS_PARM_INVALID;
+
+  wbuf = create_pdu_wbuf(0, 0, NULL);
+  if (!wbuf)
+    return BT_STATUS_NOMEM;
+
+  status = bthh_interface->get_report(&bd_addr, reportType, reportId,
+    bufferSize);
+  if (status != BT_STATUS_SUCCESS)
+    goto err_bthh_interface_get_report;
+
+  init_pdu(&wbuf->buf.pdu, cmd->service, cmd->opcode);
+  send_pdu(wbuf);
+
+  return BT_STATUS_SUCCESS;
+err_bthh_interface_get_report:
+  cleanup_pdu_wbuf(wbuf);
+  return status;
+}
+
+static bt_status_t
+opcode_set_report(const struct pdu* cmd)
+{
+  long off;
+  bt_bdaddr_t bd_addr;
+  bthh_report_type_t reportType;
+  char* report;
+  struct pdu_wbuf* wbuf;
+  bt_status_t status;
+
+  assert(bthh_interface);
+  assert(bthh_interface->set_report);
+
+  report = NULL;
+
+  off = read_bt_bdaddr_t(cmd, 0, &bd_addr);
+  if (off < 0) {
+    status = BT_STATUS_PARM_INVALID;
+    goto err_read_bt_bdaddr_t;
+  }
+
+  off = read_pdu_at(cmd, off, "C", &reportType);
+  if (off < 0) {
+    status = BT_STATUS_PARM_INVALID;
+    goto err_read_pdu_at;
+  }
+
+  off = read_pdu_at(cmd, off, "0", &report);
+  if (off < 0) {
+    status = BT_STATUS_PARM_INVALID;
+    goto err_read_pdu_at;
+  }
+
+  wbuf = create_pdu_wbuf(0, 0, NULL);
+  if (!wbuf) {
+    status = BT_STATUS_NOMEM;
+    goto err_create_pdu_wbuf;
+  }
+
+  status = bthh_interface->set_report(&bd_addr, reportType, report);
+  if (status != BT_STATUS_SUCCESS)
+    goto err_bthh_interface_set_report;
+
+  init_pdu(&wbuf->buf.pdu, cmd->service, cmd->opcode);
+  send_pdu(wbuf);
+
+  return BT_STATUS_SUCCESS;
+err_bthh_interface_set_report:
+  cleanup_pdu_wbuf(wbuf);
+err_read_bt_bdaddr_t:
+err_read_pdu_at:
+err_create_pdu_wbuf:
+  free(report);
+  return status;
+}
+
+static bt_status_t
+opcode_send_data(const struct pdu* cmd)
+{
+  long off;
+  bt_bdaddr_t bd_addr;
+  char* data;
+  struct pdu_wbuf* wbuf;
+  bt_status_t status;
+
+  assert(bthh_interface);
+  assert(bthh_interface->send_data);
+
+  data = NULL;
+
+  off = read_bt_bdaddr_t(cmd, 0, &bd_addr);
+  if (off < 0) {
+    status = BT_STATUS_PARM_INVALID;
+    goto err_read_bt_bdaddr_t;
+  }
+
+  off = read_pdu_at(cmd, off, "0", &data);
```

Comment:
Like in the previous function, the length is given in a 2-byte field before the data.